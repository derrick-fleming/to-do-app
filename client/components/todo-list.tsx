import React, { useEffect, useState } from "react";
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import { useDispatch, useSelector } from "react-redux";
import { getTodosAsync, toggleComplete, deleteTodo } from '../redux/todos-slice';
import Form from 'react-bootstrap/Form';
import FilterTodos from "./filter-component";

type Todos = {
  task: string,
  isCompleted: boolean,
  todoId: string
};

export function TodoItem(props: { todo: Todos, filter: boolean }) {
  const { todoId, task, isCompleted } = props.todo;
  const { filter } = props;
  const dispatch = useDispatch();

  const idAttr = `todo-item-${todoId}`;
  const taskClass = isCompleted === filter
    ? ''
    : 'hide-results';

  const handleChange = async (todoId: string) => {
    try {
      const complete = isCompleted ? false : true;
      const statusObject = { todoId, isCompleted: complete}
      const body = {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(statusObject)
        }
      const response = await fetch ('api/todos', body);
      const update = await response.json();
      if (update) {
        // @ts-ignore
        dispatch(toggleComplete(update));
      }
    } catch (err) {
      console.error('There was an error!', err);
    }
  }

  const handleDelete = async (todoId: string) => {
    try {
      const body = {
        method: 'DELETE',
        headers: { 'Content-Type': 'application/json'},
        body: JSON.stringify({todoId})
      };
      const response = await fetch ('api/todos', body);
      const deleteStatus = await response.json();
      if (deleteStatus) {
        // @ts-ignore
        dispatch(deleteTodo(todoId));
      }
    } catch (err) {
      console.error('There was an error!', err);
    }
  }


    return (
      <li className={`${taskClass}`}>
      <div>
        <Form.Check type="checkbox" id={idAttr} label={task} className={` width-80 fs-4 py-2 px-2 border-bottom border-2 border-light d-inline-block`} onChange={() => handleChange(todoId)}></Form.Check>
        <button className="d-inline border-0 background-none" onClick={() => handleDelete(todoId)}><i className="fa-solid fa-trash"></i></button>
      </div>
    </li>
  )
}


export default function TodoList() {
  const [ filterComplete, setFilterComplete ] = useState(false);
  const todos = useSelector( (state:{ todosList: Todos[]}) => state.todosList);
  const dispatch = useDispatch();

  function handleFilter (id: string) {
    if (id === "complete") setFilterComplete(true);
    if (id === "incomplete") setFilterComplete(false);
  }

  useEffect (() => {
    // @ts-ignore
    dispatch(getTodosAsync());
  }, [dispatch]);

  const list = todos.length === 0 ? 'Loading...' :
    todos.map(todo => {
      return (
        <TodoItem
        key={todo.todoId}
        todo={todo}
        filter={filterComplete} />
      )
    });

  return (
    <Row className="justify-content-center">
      <Col md={8} xs={11}>
        <FilterTodos filter={handleFilter}/>
      </Col>
      <Col md={8} xs={11}>
        <ul>
          {list}
        </ul>
      </Col>
    </Row>
  )
}
