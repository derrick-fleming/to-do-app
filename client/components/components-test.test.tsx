import React from 'react';
import renderer from 'react-test-renderer';
import TodoForm from './todo-form';
import TodoList from './todo-list';
import { TodoItem } from './todo-list';
import { Provider } from 'react-redux';
import store from '../redux/generate-store';

it('renders TodoList correctly', () => {
  const tree = renderer
    .create(<Provider store={store}><TodoList /></Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders TodoForm correctly', () => {
  const tree = renderer
    .create(<Provider store={store}><TodoForm /></Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders TodoItem correctly', () => {
  const component = renderer.create(<Provider store={store}><TodoItem key='1' todo={{ todoId: '1', task: 'Complete work', isCompleted: false }} /></Provider>)
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
