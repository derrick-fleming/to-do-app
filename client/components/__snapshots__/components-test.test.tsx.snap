// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TodoForm correctly 1`] = `
<div
  className="justify-content-center row"
>
  <div
    className="col-md-8 col-11"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <label
        className="fs-4 form-label"
        htmlFor="todo"
      >
        Write your Task
      </label>
      <input
        className="my-2 form-control"
        id="todo"
        onChange={[Function]}
        placeholder="Enter task"
        type="text"
        value=""
      />
      <div
        className="text-end"
      >
        <button
          className="fs-4 btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`renders TodoItem correctly 1`] = `
<li>
  <div>
    <div
      className=" fs-4 py-2 px-2 border-bottom border-2 border-light form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        id="todo-item-1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="todo-item-1"
        title=""
      >
        Complete work
      </label>
    </div>
  </div>
</li>
`;

exports[`renders TodoList correctly 1`] = `
<div
  className="justify-content-center row"
>
  <div
    className="col-md-8 col-11"
  >
    <ul>
      Loading...
    </ul>
  </div>
</div>
`;
